File: src/ast/FunctionType.java
package ast;

import java.io.StringWriter;

public class FunctionType extends BaseNode {
    private BaseNode _returnType;
    private BaseNode _params;
    private BaseNode _cvQualifier;
    private SimpleReferenceType _referenceQualifier;
    private BaseNode _exceptionSpec;

    public FunctionType(BaseNode returnType, BaseNode Params, BaseNode cvQualifier, SimpleReferenceType referenceQualifier, BaseNode exceptionSpec) {
        super(NodeType.FunctionType);
        _returnType = returnType;
        _params = Params;
        _cvQualifier = cvQualifier;
        _referenceQualifier = referenceQualifier;
        _exceptionSpec = exceptionSpec;
    }

    @Override
    public void PrintLeft(StringWriter writer) {
        _returnType.PrintLeft(writer);
        writer.write(" ");
    }

    @Override
    public void PrintRight(StringWriter writer) {
        writer.write("(");
        _params.Print(writer);
        writer.write(")");

        _returnType.PrintRight(writer);

        _cvQualifier.Print(writer);

        if (_referenceQualifier.Qualifier != Reference.None) {
            writer.write(" ");
            _referenceQualifier.PrintQualifier(writer);
        }

        if (_exceptionSpec != null) {
            writer.write(" ");
            _exceptionSpec.Print(writer);
        }
    }

    @Override
    public boolean HasRightPart() {
        return true;
    }

    @Override
    public boolean HasFunctions() {
        return true;
    }
}

File: src/de/fearlesstobi/demangler/ast/FunctionType.kt
package de.fearlesstobi.demangler.ast

import java.io.StringWriter

class FunctionType(private val returnType: BaseNode, private val params: BaseNode, private val cvQualifier: BaseNode, private val referenceQualifier: SimpleReferenceType, private val exceptionSpec: BaseNode?) : BaseNode(NodeType.FunctionType) {
    public override fun printLeft(writer: StringWriter) {
        returnType.printLeft(writer)
        writer.write(" ")
    }

    public override fun printRight(writer: StringWriter) {
        writer.write("(")
        params.print(writer)
        writer.write(")")
        returnType.printRight(writer)
        cvQualifier.print(writer)
        if (referenceQualifier.qualifier != Reference.None) {
            writer.write(" ")
            referenceQualifier.PrintQualifier(writer)
        }
        if (exceptionSpec != null) {
            writer.write(" ")
            exceptionSpec.print(writer)
        }
    }

    public override fun hasRightPart(): Boolean {
        return true
    }

    public override fun hasFunctions(): Boolean {
        return true
    }

}